### PR TITLE
Add logo style badges

### PR DESCRIPTION
**Pull request by RedSparr0w**
This PR adds 2 new style badges which emphasize the logo on the badge:
- `popout`
- `popout-square`

These allow for a query parameter called `logoPosition`:
_(By default the logo is centered)_
negative numbers moves the badge down/logo up:
`-5`:

positive numbers move the badge up/logo down:
`5`:

If a logo is not specified it will automatically fallback to serve the `flat` or `flat-square` style badge

### Examples:

`popout` | `popout-square`
:--: | :--:
image | image
image | image
image | image
image | image
image | image
image | image
image | image
image | image
image | image
image | image
image | image

### CODE REVIEW

**Comment by paulmelnikow** [comment]
Wow, neat, these are really cute!

I wonder if anyone else would like to weigh in on the design. Did you experiment with the logo centered over the bar instead of at the top? I wonder how that would look.

File: lib/supported-features.js [hunk 1]
```diff
@@ -5,6 +5,8 @@ const advertisedStyles = [
   'flat',
   'flat-square',
   'for-the-badge',
+  'flat-logo',
+  'flat-square-logo',
```
Comment (paulmelnikow) [inline lib/supported-features.js]: Do you think `popout` and `popout-square` might be clearer names for these formats?

Comment (RedSparr0w) [reply lib/supported-features.js]: Yeah good idea, i couldn't think of what to call them, but `popout` is a great choice.

**Comment by RedSparr0w** [comment]
>Did you experiment with the logo centered over the bar instead of at the top?

Never thought of that,
Added a few example badges with centered logo,
Not really sure which ones i prefer now though :laughing:

**Comment by PyvesB** [comment]
Not sure which ones I prefer either. When it comes to objects or geometric logos, the centered ones look better in my opinion,  but for instance you get the impression that the Linux penguin is "sitting"  on the badge, it's quite quite a neat effect as well. Possibly we could give users two options `popout` and `popout-centered` (and the square variants of these two)?

**Comment by bkdotcom** [comment]
They are pretty, but....

seems to break the non-promotional mission statement of the project
https://github.com/badges/shields/blob/master/spec/SPECIFICATION.md

**Comment by paulmelnikow** [comment]
> Not sure which ones I prefer either. When it comes to objects or geometric logos, the centered ones look better in my opinion,  but for instance you get the impression that the Linux penguin is "sitting"  on the badge, it's quite quite a neat effect as well. Possibly we could give users two options `popout` and `popout-centered` (and the square variants of these two)?

I like the penguin sitting on the floor! For all the rest I prefer the centered. Let's definitely support that one.

Actually, I'd be curious to see what the penguin looks like centered. If it looks decent I would say, let's support centered and not floored. If it looks bad, and you're up for doing the work, we could support both logo placement options.

It might be better for the user to choose between them using a parameter like `?logoAlign=bottom`.

> seems to break the non-promotional mission statement of the project

That's true of all the logos. I don't think we're veering so far afield by supporting these.

**Comment by RedSparr0w** [comment]
Or what about if i did default centered, and had `?logoAlign` as an int:
`?logoAlign=10`
image

`?logoAlign=6`
image

`?logoAlign=-10`
image

`?logoAlign=-7`
image

_possibly swap around the positive/negative, not sure which makes more sense_

**Comment by paulmelnikow** [comment]
How about `logoTop=10` moving it down, `logoTop=-10` moving it up?

Or `logoPosition=10`/`logoPosition=-10`?

Penguin looks cuter at `=10` or `=6` than centered!

**Comment by RedSparr0w** [comment]
`logoPosition` sounds good.

I almost think the param shouldn't include the word logo at all, as the logo won't move, just the badge does.
Possibly something like `popoutPosition` as it will _currently_ only be used for the `popout` style badges,
But probably better to future proof it with a more generic parameter.

Definitely looks a lot better with the badge underneath! :smile: 

I also need to update the example badge to use a logo, perhaps the shields favicon? 
_(once it's updated to be a little higher res)_
image

**Comment by paulmelnikow** [comment]
> perhaps the shields favicon

I like your examples better!

While technically it's true that the badge is moving, I don't think people will necessarily see it that way. Since it goes with the `logo=` parameter I think it's good to be named with that. Also it's similar, in a way, to `logoWidth=`. I think `popoutPosition` could work too, though you're probably right that the position parameter could potentially apply to other badges someday.

**Comment by RedSparr0w** [comment]
>I like your examples better!

Thanks :smile:
Tried _a shields logo_, and it blends into the badge way too much aswell:


Have updated to use `logoPosition`
negative numbers moves the badge up/logo down:
`-5`:

positive numbers move the badge down/logo up:
`5`:


Still not sure what logo to use in the examples as all the logos in the folder don't fit too well:
(lighter logos on the white background, darker logos on the grey of the badge)





Think the twitter or bitcoin probably look best from the available logos.

**Comment by paulmelnikow** [comment]
That sign for `logoPosition` feels flipped from what I'd expect. Most 2D canvases put the origin at the top left and use positive numbers to mean down – including CSS, HTML5 Canvas, and SVG – and as I mentioned I think people will think of the logo as the thing that's moving, not the badge. That's pure intuition, though, so if you feel strongly I'll defer.

> Think the twitter or bitcoin probably look best from the available logos.

Agreed!

**Comment by RedSparr0w** [comment]
I've reverse `logoPosition`, but I am unsure how to get the logo in the styles example.

**Comment by paulmelnikow** [comment]
Making this change in `renderStyleExamples()` in `usage.js` should do the trick:

```js
const logo = style.startsWith('popout') ? 'data:something-cool-here' : undefined;
const badgeUri = staticBadgeUrl(
  baseUri,
  'style',
  style,
  'green',
  { longCache, style, logo });
```

I'm wondering though, if we should add a separate heading, and show a few examples showing logos, so people are seeing e.g. flag-square, social, and also popout. That way, people are less likely to conflate _a badge with a logo_ with _popout_.

**Comment by paulmelnikow** [comment]
We should get this merged! Is it just waiting for usage instructions to be updated?

**Comment by paulmelnikow** [comment]
If it is just usage, let's merge it, and then add the docs in a follow-on PR.

**Comment by RedSparr0w** [comment]
All working with latest changes merged.
Could you give it a quick look over before merge @paulmelnikow

**Comment by olivierlacan** [comment]
Do we really seriously honestly need to keep adding exotic styles like this? 

The inbalance in the design due to the logos breaking out of the container means that these are going to be a mess on any page that includes more than one of them. 

This project is not an emoji or trading card collection, it's a serious attempt at meaningful standardization of metadata for open source projects. It's a regression to the exact issue that [prompted me to create Shields in the first place](https://olivierlacan.com/posts/an-open-source-rage-diamond/): inconsistency, visual noise for no other purpose than being louder than other badges, all in service of a superficial aesthetic pursuit that frankly looks shoddy most of the time.

Sorry if this feels like a strong diss, and I know the project is in your care now @paulmelnikow but at the risk of looking like I'm swooping in after everyone's done with the hard work, I wish we could put our collective foot down on these alternative designs.

Paralysis of choice is a real problem. Offering a million design alternatives means that we're destroying the useful standardization we've worked so hard to achieve for years. Let's be brave and deprecate designs if we believe we've found a better one, but at this cadence we'll have 30 different styles in a few years and we'll be back in a visual mess where badges are shouting against each other for attention rather than showcasing useful information at a glance in a succinct a clear manner.

**Comment by olivierlacan** [comment]
@paulmelnikow I'm so glad you're shepherding this project. Apologies to @RedSparr0w for the drive-by cranky post. I realize I'm singling you out to make a point and that can't feel fair when you've worked your ass off. 

The last thing I want is to add some gatekeeping antics to this healthy blossoming community. I love your suggestion for best practices @paulmelnikow. This is basically what I moved toward with https://keepachangelog.com/. I'm really hopeful the generator ideas related to the #1899 redesign will facilitate this, since it would make sense to guide people through the process.

I could swear there was an old discussion somewhere around changing the default to flat when flat design started becoming real trendy. Yep, [there is is](https://github.com/badges/shields/issues/185#issuecomment-44990248).

It's my job to make sure the original rationale behind Shields is clear and documented. I think we had it in the original repo and it might have gotten lost in the weeds of time.

Here are some fun historical items I managed to dig back up:
- https://github.com/badges/shields/blob/master/spec/SPECIFICATION.md
- https://github.com/badges/shields/blob/historical/static/shields_white%402x.png
- https://github.com/badges/shields/blob/historical/static/proportions.png
- https://github.com/badges/shields/tree/historical#guidelines
- https://github.com/badges/shields/tree/historical#examples

**Comment by RedSparr0w** [comment]
No worries, I definitely understand your point of view, and if we decide to downsize the current styles i'm fine with that.

> It would be helpful to write some design guidelines for the project, both to guide developers in maintaining consistent badge URLs, and to shaping the way we deal with styles. A guideline is a good way to put down our collective foot.

:+1:

>I’ve been meaning to propose changing the default to flat-square, and perhaps dropping plastic

It would be great if we had some metrics on the styles of badges being used.

>The inbalance in the design due to the logos breaking out of the container means that these are going to be a mess on any page that includes more than one of them.

<details>
<summary>The main kind of use case I imagined with this badge would have been something similar to:</summary>

# My Project
  
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed viverra non eros sit amet rutrum. Curabitur ut dictum mi, a pulvinar odio. Praesent rhoncus justo augue, at euismod felis porttitor eu. Sed auctor leo sit amet lobortis elementum. Fusce eu nulla sit amet dui rhoncus cursus. Ut accumsan, urna ut.

## Downloads:
  

## Something Else
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse non sem enim. Suspendisse vitae nisl aliquet, laoreet erat quis, blandit turpis. Mauris vel luctus diam. Duis neque quam, ornare vel aliquam et, tristique quis risus. In tincidunt enim a sapien suscipit, ut malesuada tortor iaculis. Proin cursus, tortor vel commodo.

## Donate:
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam in pellentesque est. Nullam elementum leo at tempor dictum. Donec commodo commodo mauris ut laoreet. Curabitur.


# -END-
<hr/>
</details>
<br/>

Similarly with the [`for-the-badge` style](https://forthebadge.com/), If the user is already going to be using these i feel it would be better to allow them to match our badges with those to maintain some degree of consistency.

I guess the thing to decide now is where we go from here,
It may be worth opening an issue so we can keep track of this for future reference and gather some more discussion.

**Pull request merged**